oke/oke_manager_client.go

```diff
@@ -0,0 +1,1398 @@
+func (mgr *ClusterManagerClient) DeleteVCN(ctx context.Context, vcnID string) error {
+
+	logrus.Debugf("deleting VCN with VCN ID %s", vcnID)
+
+	if len(vcnID) == 0 {
+		return fmt.Errorf("vcnID must be set to delete the VCN")
+	}
+
+	getVCNReq := core.GetVcnRequest{}
+	getVCNReq.VcnId = common.String(vcnID)
+
+	getVCNResp, err := mgr.virtualNetworkClient.GetVcn(ctx, getVCNReq)
+	if err != nil {
+		logrus.Debugf("get VCN failed with err %v", err)
+		return err
+	}
+
+	// The VCN must be completely empty (meaning no subnets, attached gateways, or security lists)
+
+	// Delete Route Rules
+	listRouteTblsReq := core.ListRouteTablesRequest{}
+	listRouteTblsReq.VcnId = common.String(vcnID)
+	listRouteTblsReq.CompartmentId = getVCNResp.Vcn.CompartmentId
+	rtResp, err := mgr.virtualNetworkClient.ListRouteTables(ctx, listRouteTblsReq)
+	if err != nil {
+		logrus.Debugf("list route tables failed with err %v", err)
+		return err
+	}
+	for _, rt := range rtResp.Items {
+
+		updateRTReq := core.UpdateRouteTableRequest{}
+		updateRTReq.RtId = rt.Id
+		updateRTReq.RouteRules = []core.RouteRule{}
+
+		logrus.Debugf("removing default route rule from route table %s", *rt.Id)
+		_, err = mgr.virtualNetworkClient.UpdateRouteTable(ctx, updateRTReq)
+		if err != nil {
+			logrus.Debugf("update route table failed with err %v", err)
+			return err
+		}
+	}
+
+	// Delete Internet Gateways from VCN
+	listIGsReq := core.ListInternetGatewaysRequest{}
+	listIGsReq.VcnId = common.String(vcnID)
+	listIGsReq.CompartmentId = getVCNResp.Vcn.CompartmentId
+	igsResp, err := mgr.virtualNetworkClient.ListInternetGateways(ctx, listIGsReq)
+	if err != nil {
+		logrus.Debugf("list internet gateway(s) failed with err %v", err)
+		return err
+	}
+	for _, ig := range igsResp.Items {
+		deleteIGReq := core.DeleteInternetGatewayRequest{}
+		deleteIGReq.IgId = ig.Id
+		logrus.Debugf("deleting internet gateway %s", *ig.Id)
+		_, err = mgr.virtualNetworkClient.DeleteInternetGateway(ctx, deleteIGReq)
+		if err != nil {
+			logrus.Debugf("warning: delete internet gateway failed with err %v", err)
+			// Continue tearing down.
+		}
+	}
+
+	// Delete all the subnets from VCN
+	listSubnetsReq := core.ListSubnetsRequest{}
+	listSubnetsReq.VcnId = common.String(vcnID)
+	listSubnetsReq.CompartmentId = getVCNResp.Vcn.CompartmentId
+	listSubnetResp, err := mgr.virtualNetworkClient.ListSubnets(ctx, listSubnetsReq)
+	if err != nil {
+		logrus.Debugf("list subnets failed with err %v", err)
+		return err
+	}
+	for _, subnet := range listSubnetResp.Items {
+		deleteSubnetReq := core.DeleteSubnetRequest{}
+		deleteSubnetReq.SubnetId = subnet.Id
+		logrus.Debugf("deleting subnet %s", *subnet.Id)
+		_, err := mgr.virtualNetworkClient.DeleteSubnet(ctx, deleteSubnetReq)
+		if err != nil {
+			logrus.Debugf("warning: delete subnet failed with err %v", err)
+			// Continue tearing down.
+		}
+	}
+	// TODO better to poll instead of sleep
+	time.Sleep(mgr.sleepDuration * time.Second)
```

Comment:
I would guess that this is unreliable

Reply:
It's been reliable enough to leave this as a TODO / enhancement for a subsequent commit.